File: go/parser/parser_test.go
package parser

import (
	"fmt"
	"interp/ast"
	"interp/lexer"
	"testing"
)

func TestLetStatements(t *testing.T) {
	input := `
let x = 5;
let y = 10;
let foobar = 838383;
`
	l := lexer.New(input)
	p := New(l)

	program := p.ParseProgram()
	checkParserErrors(t, p)
	if program == nil {
		t.Fatalf("ParseProgram() returned nil")
	}

	if len(program.Statements) != 3 {
		t.Fatalf("program.Statements does not contain 3 statements. got=%d", len(program.Statements))
	}

	tests := []struct {
		expectedIdentifier string
	}{
		{"x"},
		{"y"},
		{"foobar"},
	}

	for i, tt := range tests {
		stmnt := program.Statements[i]
		if !testLetStatement(t, stmnt, tt.expectedIdentifier) {
			return
		}
	}
}

func TestReturnStatements(t *testing.T) {
	input := `
return 5;
return 10;
return 838383;
	`

	l := lexer.New(input)
	p := New(l)

	prog := p.ParseProgram()
	checkParserErrors(t, p)
	if prog == nil {
		t.Fatal("ParseProgram return nil")
	}

	if len(prog.Statements) != 3 {
		t.Fatalf("Program has %d statements instead of 3", len(prog.Statements))
	}

	for _, stmnt := range prog.Statements {
		r, ok := stmnt.(*ast.ReturnStatement)
		if !ok {
			t.Errorf("Statement isn't *ast.ReturnStatement, got=%T", stmnt)
			continue
		}
		if r.TokenLiteral() != "return" {
			t.Errorf("r.TokenLiteral not 'return', got %q",
				r.TokenLiteral())
		}
	}
}

func TestIdentifierExpression(t *testing.T) {
	input := `foobar;`

	l := lexer.New(input)
	p := New(l)

	prog := p.ParseProgram()
	checkParserErrors(t, p)
	if prog == nil {
		t.Fatal("ParseProgram returned nil")
	}

	if len(prog.Statements) != 1 {
		t.Errorf("Program has %d statments instead of 1", len(prog.Statements))
	}

	stmnt, ok := prog.Statements[0].(*ast.ExpressionStatement)
	if !ok {
		t.Fatalf("Statment is not an *ast.ExpressionStatement, got %T", prog.Statements[0])
	}

	ident, ok := stmnt.Expression.(*ast.Identifier)
	if !ok {
		t.Fatalf("Expression is not an *ast.Identifier, got %T", stmnt.Expression)
	}

	if ident.Value != "foobar" {
		t.Errorf("Value is not %s. got %s", "foobar", ident.Value)
	}
	if ident.TokenLiteral() != "foobar" {
		t.Errorf("TokenLiteral not %s. got %s", "foobar", ident.TokenLiteral())
	}
}

func TestIntegerLiteral(t *testing.T) {
	input := `5;`

	l := lexer.New(input)
	p := New(l)

	prog := p.ParseProgram()
	checkParserErrors(t, p)
	if prog == nil {
		t.Fatal("ParseProgram returned nil")
	}

	if len(prog.Statements) != 1 {
		t.Errorf("Program has %d statments instead of 1", len(prog.Statements))
	}

	stmnt, ok := prog.Statements[0].(*ast.ExpressionStatement)
	if !ok {
		t.Fatalf("Statment is not an *ast.ExpressionStatement, got %T", prog.Statements[0])
	}

	lit, ok := stmnt.Expression.(*ast.IntegerLiteral)
	if !ok {
		t.Fatalf("Expression is not an *ast.IntegerLiteral, got %T", stmnt.Expression)
	}

	if lit.Value != 5 {
		t.Errorf("Value is not %d. got %d", 5, lit.Value)
	}
	if lit.TokenLiteral() != "5" {
		t.Errorf("TokenLiteral not %s. got %s", "5", lit.TokenLiteral())
	}
}

func TestPrefixExpression(t *testing.T) {
	prefixTests := []struct {
		input        string
		operator     string
		integerValue int64
	}{
		{"!5", "!", 5},
		{"-15", "-", 15},
	}

	for _, tt := range prefixTests {
		l := lexer.New(tt.input)
		p := New(l)

		prog := p.ParseProgram()
		checkParserErrors(t, p)
		if prog == nil {
			t.Error("ParseProgram returned nil")
			continue
		}

		if len(prog.Statements) != 1 {
			t.Errorf("there is not 1 Statment, got %d", len(prog.Statements))
			continue
		}

		stmnt, ok := prog.Statements[0].(*ast.ExpressionStatement)
		if !ok {
			t.Errorf("stmnt was not *ast.ExpressionStatment, got %T", prog.Statements[0])
			continue
		}

		pre, ok := stmnt.Expression.(*ast.PrefixExpression)
		if !ok {
			t.Errorf("expression was not *ast.PrefixExpression, got %T", stmnt.Expression)
			continue
		}

		if pre.Operator != tt.operator {
			t.Errorf("operator was %s instead of %s", pre.Operator, tt.operator)
		}

		testIntegerLiteral(t, pre.Right, tt.integerValue)
	}
}

func testLetStatement(t *testing.T, s ast.Statement, name string) bool {
	if s.TokenLiteral() != "let" {
		t.Errorf("s.TokenLiteral not 'let'. got=%q", s.TokenLiteral())
		return false
	}
	letStmt, ok := s.(*ast.LetStatement)
	if !ok {
		t.Errorf("s not *ast.LetStatement. got=%T", s)
		return false
	}
	if letStmt.Name.Value != name {
		t.Errorf("letStmt.Name.Value not '%s'. got=%s", name, letStmt.Name.Value)
		return false
	}
	if letStmt.Name.TokenLiteral() != name {
		t.Errorf("s.Name not '%s'. got=%s", name, letStmt.Name)
		return false
	}
	return true
}

func checkParserErrors(t *testing.T, p *Parser) {
	errors := p.Errors()
	if len(errors) == 0 {
		return
	}
	t.Errorf("parser has %d errors", len(errors))
	for _, msg := range errors {
		t.Errorf("parser error: %q", msg)
	}
	t.FailNow()
}

func testIntegerLiteral(t *testing.T, il ast.Expression, value int64) bool {
	integ, ok := il.(*ast.IntegerLiteral)
	if !ok {
		t.Errorf("il not *ast.IntegerLiteral. got=%T", il)
		return false
	}
	if integ.Value != value {
		t.Errorf("integ.Value not %d. got=%d", value, integ.Value)
		return false
	}
	if integ.TokenLiteral() != fmt.Sprintf("%d", value) {
		t.Errorf("integ.TokenLiteral not %d. got=%s", value,
			integ.TokenLiteral())
		return false
	}
	return true
}

File: cpp/src/parser/parser_test.cpp
#include <gtest/gtest.h>
#include <tuple>

#include "parser.h"

void check_parser_errors(interp::parser::Parser p);
void test_let_statment(std::shared_ptr<interp::ast::Statement> stmnt, std::string name);
void test_integer_literal(std::shared_ptr<interp::ast::Expression> expr, int64_t value);
void test_identifier(std::shared_ptr<interp::ast::Expression> expr, std::string value);

TEST(ParserTest, TestLetStatment)
{
	std::string input = R"(
let x = 5;
let y = 10;
let foobar = 838383;
	)";

	auto lex = interp::lexer::Lexer(input);
	auto parse = interp::parser::Parser(lex);

	auto prog = parse.parse_program();
	check_parser_errors(parse);

	EXPECT_EQ(3, prog->statements.size())
		<< "program.Statements does not contain 3 statements. got=" << std::to_string(prog->statements.size());

	std::vector<std::string> expected = {"x", "y", "foobar"};

	for (size_t i = 0; i < expected.size(); i++)
	{
		auto stmnt = prog->statements.at(i);
		test_let_statment(stmnt, expected[i]);
	}
}

TEST(ParserTest, TestReturnStatment)
{
	std::string input = R"(
return 5;
return 10;
return 838383;
	)";

	interp::lexer::Lexer lex(input);
	interp::parser::Parser parse(lex);

	auto prog = parse.parse_program();
	check_parser_errors(parse);

	EXPECT_EQ(3, prog->statements.size())
		<< "program.Statements does not contain 3 statements. got=" << std::to_string(prog->statements.size());

	for (auto stmnt : prog->statements)
	{
		if (interp::ast::ReturnStatement *retstmnt = dynamic_cast<interp::ast::ReturnStatement *>(stmnt.get()))
		{
			EXPECT_EQ("return", retstmnt->token_literal()) << "token literal not return got " << retstmnt->token_literal();
		}
		else
		{
			EXPECT_TRUE(false) << "stmnt not ReturnStatement";
		}
	}
}

TEST(ParserTest, TestIdentifierExpression)
{
	std::string input = "foobar;";

	interp::lexer::Lexer lex(input);
	interp::parser::Parser parse(lex);

	auto prog = parse.parse_program();
	check_parser_errors(parse);

	ASSERT_EQ(1, prog->statements.size())
		<< "program.Statements does not contain 1 statements. got=" << std::to_string(prog->statements.size());

	if (interp::ast::ExpressionStatement *expstmnt = dynamic_cast<interp::ast::ExpressionStatement *>(prog->statements[0].get()))
	{
		if (interp::ast::Identifier *ident = dynamic_cast<interp::ast::Identifier *>(expstmnt->expression.get()))
		{
			EXPECT_EQ("foobar", ident->value) << "value not foobar got " << ident->value;
			EXPECT_EQ("foobar", ident->token_literal()) << "token literal not foobar got " << ident->token_literal();
		}
		else
		{
			EXPECT_TRUE(false) << "expression not Identifier";
		}
	}
	else
	{
		EXPECT_TRUE(false) << "stmnt not ExpressionStatement";
	}
}

TEST(ParserTest, TestIntegerLiteral)
{
	std::string input = "5;";

	interp::lexer::Lexer lex(input);
	interp::parser::Parser parse(lex);

	auto prog = parse.parse_program();
	check_parser_errors(parse);

	ASSERT_EQ(1, prog->statements.size())
		<< "program.Statements does not contain 1 statements. got=" << std::to_string(prog->statements.size());

	if (interp::ast::ExpressionStatement *expstmnt = dynamic_cast<interp::ast::ExpressionStatement *>(prog->statements[0].get()))
	{
		if (interp::ast::IntegerLiteral *intlit = dynamic_cast<interp::ast::IntegerLiteral *>(expstmnt->expression.get()))
		{
			EXPECT_EQ(5LL, intlit->value) << "value not 5 got " << intlit->value;
			EXPECT_EQ("5", intlit->token_literal()) << "token literal not 5 got " << intlit->token_literal();
		}
		else
		{
			EXPECT_TRUE(false) << "expression not IntegerLiteral";
		}
	}
	else
	{
		EXPECT_TRUE(false) << "stmnt not ExpressionStatement";
	}
}

TEST(ParserTest, TestBooleanExpression)
{
	std::string input = "true;";

	interp::lexer::Lexer lex(input);
	interp::parser::Parser parse(lex);

	auto prog = parse.parse_program();
	check_parser_errors(parse);

	ASSERT_EQ(1, prog->statements.size())
		<< "program.Statements does not contain 1 statements. got=" << std::to_string(prog->statements.size());

	if (interp::ast::ExpressionStatement* expstmnt = dynamic_cast<interp::ast::ExpressionStatement*>(prog->statements[0].get()))
	{
		if (interp::ast::Boolean* ident = dynamic_cast<interp::ast::Boolean*>(expstmnt->expression.get()))
		{
			EXPECT_EQ(true, ident->value) << "value not true got " << ident->value;
			EXPECT_EQ("true", ident->token_literal()) << "token literal not true got " << ident->token_literal();
		}
		else
		{
			EXPECT_TRUE(false) << "expression not Boolean";
		}
	}
	else
	{
		EXPECT_TRUE(false) << "stmnt not ExpressionStatement";
	}
}

TEST(ParserTest, TestPrefixExpression)
{
	std::tuple<std::string, std::string, int64_t> expected[] = {
		std::tuple("!5", "!", 5),
		std::tuple("-15", "-", 15),
	};

	for (auto tt : expected)
	{

		interp::lexer::Lexer lex(std::get<0>(tt));
		interp::parser::Parser parse(lex);

		auto prog = parse.parse_program();
		check_parser_errors(parse);

		ASSERT_EQ(1, prog->statements.size())
			<< "program.Statements does not contain 1 statements. got=" << std::to_string(prog->statements.size());

		if (interp::ast::ExpressionStatement *expstmnt = dynamic_cast<interp::ast::ExpressionStatement *>(prog->statements[0].get()))
		{
			if (interp::ast::PrefixExpression *pref = dynamic_cast<interp::ast::PrefixExpression *>(expstmnt->expression.get()))
			{

				EXPECT_EQ(std::get<1>(tt), pref->p_operator) << "operatior not " << std::get<1>(tt) << " got " << pref->p_operator;

				test_integer_literal(pref->right, std::get<2>(tt));
			}
			else
			{
				EXPECT_TRUE(false) << "expression not PrefixExpression";
			}
		}
		else
		{
			EXPECT_TRUE(false) << "stmnt not ExpressionStatement";
		}
	}
}

TEST(ParserTest, TestInfixExpression)
{
	std::tuple<std::string, int64_t, std::string, int64_t> expected[] = {
		std::tuple("5 + 5;", 5, "+", 5),
		std::tuple("5 - 5;", 5, "-", 5),
		std::tuple("5 * 5;", 5, "*", 5),
		std::tuple("5 / 5;", 5, "/", 5),
		std::tuple("5 >= 5;", 5, ">=", 5),
		std::tuple("5 <= 5;", 5, "<=", 5),
		std::tuple("5 > 5;", 5, ">", 5),
		std::tuple("5 < 5;", 5, "<", 5),
		std::tuple("5 == 5;", 5, "==", 5),
		std::tuple("5 != 5;", 5, "!=", 5),
	};

	for (auto tt : expected)
	{

		interp::lexer::Lexer lex(std::get<0>(tt));
		interp::parser::Parser parse(lex);

		auto prog = parse.parse_program();
		check_parser_errors(parse);

		ASSERT_EQ(1, prog->statements.size())
			<< "program.Statements does not contain 1 statements. got=" << std::to_string(prog->statements.size());

		if (interp::ast::ExpressionStatement* expstmnt = dynamic_cast<interp::ast::ExpressionStatement*>(prog->statements[0].get()))
		{
			if (interp::ast::InfixExpression* pref = dynamic_cast<interp::ast::InfixExpression*>(expstmnt->expression.get()))
			{
				test_integer_literal(pref->left, std::get<1>(tt));

				EXPECT_EQ(std::get<2>(tt), pref->p_operator) << "operatior not " << std::get<2>(tt) << " got " << pref->p_operator;

				test_integer_literal(pref->right, std::get<3>(tt));
			}
			else
			{
				EXPECT_TRUE(false) << "expression not InfixExpression";
			}
		}
		else
		{
			EXPECT_TRUE(false) << "stmnt not ExpressionStatement";
		}
	}
}

TEST(ParserTest, TestOperatorPrecedence)
{
	std::tuple<std::string, std::string> expected[] = {
		std::tuple("true", "true"),
		std::tuple("false", "false"),
		std::tuple("3 > 5 == false", "((3 > 5) == false)"),
		std::tuple("3 < 5 == true", "((3 < 5) == true)"),

		std::tuple("1 + (2 + 3) + 4", "((1 + (2 + 3)) + 4)"),
		/*std::tuple("(5 + 5) * 2", "((5 + 5) * 2)"),
		std::tuple("2 / (5 + 5)", "(2 / (5 + 5))"),
		std::tuple("-(5 + 5)", "(-(5 + 5))"),
		std::tuple("!(true == true)", "(!(true == true))"),*/
	};

	for (auto tt : expected)
	{

		interp::lexer::Lexer lex(std::get<0>(tt));
		interp::parser::Parser parse(lex);

		auto prog = parse.parse_program();
		check_parser_errors(parse);

		ASSERT_EQ(1, prog->statements.size())
			<< "program.Statements does not contain 1 statements. got=" << std::to_string(prog->statements.size());

		EXPECT_EQ(std::get<1>(tt), prog->statements[0]->string());
	}
}

void check_parser_errors(interp::parser::Parser p)
{
	auto errors = p.get_errors();
	if (errors.size() == 0)
	{
		return;
	}

	EXPECT_TRUE(false) << "parser has " << std::to_string(errors.size()) << " errors";
	for (auto error : errors)
	{
		EXPECT_TRUE(false) << "parser error: " << error;
	}
	ASSERT_TRUE(false);
}

void test_let_statment(std::shared_ptr<interp::ast::Statement> stmnt, std::string name)
{
	ASSERT_EQ("let", stmnt->token_literal()) << "stmnt literal not 'let'. got=" << stmnt->token_literal();

	if (interp::ast::LetStatement *letstmnt = dynamic_cast<interp::ast::LetStatement *>(stmnt.get()))
	{
		ASSERT_EQ(name, letstmnt->name.value) << "name not " << name << " got " << letstmnt->name.value;
		ASSERT_EQ(name, letstmnt->name.token_literal()) << "name.token_literal not " << name << " got " << letstmnt->name.token_literal();
	}
	else
	{
		ASSERT_TRUE(false) << "stmnt not LetStatement";
	}
}

void test_integer_literal(std::shared_ptr<interp::ast::Expression> expr, int64_t value)
{
	if (interp::ast::IntegerLiteral* intlit = dynamic_cast<interp::ast::IntegerLiteral*>(expr.get()))
	{

		EXPECT_EQ(value, intlit->value) << "value not " << value << " got " << intlit->value;
	}
	else
	{
		EXPECT_TRUE(false) << "expression not IntegerLiteral";
	}
}

void test_identifier(std::shared_ptr<interp::ast::Expression> expr, std::string value)
{
	if (interp::ast::Identifier* ident = dynamic_cast<interp::ast::Identifier*>(expr.get()))
	{
		EXPECT_EQ(value, ident->value) << "value not " << value << " got " << ident->value;
		EXPECT_EQ(value, ident->token_literal()) << "token literal not " << value << " got " << ident->token_literal();
	}
	else
	{
		EXPECT_TRUE(false) << "expression not Identifier";
	}
}
